cub/cub/device/device_segmented_reduce.cuh

```diff
@@ -208,35 +273,36 @@ struct DeviceSegmentedReduce
             typename EndOffsetIteratorT,
             typename ReductionOpT,
             typename T>
-  CUB_RUNTIME_FUNCTION static cudaError_t Reduce(void *d_temp_storage,
-                                                 size_t &temp_storage_bytes,
-                                                 InputIteratorT d_in,
-                                                 OutputIteratorT d_out,
-                                                 int num_segments,
-                                                 BeginOffsetIteratorT d_begin_offsets,
-                                                 EndOffsetIteratorT d_end_offsets,
-                                                 ReductionOpT reduction_op,
-                                                 T initial_value,
-                                                 cudaStream_t stream = 0)
+  CUB_RUNTIME_FUNCTION static cudaError_t Reduce(
+    void* d_temp_storage,
+    size_t& temp_storage_bytes,
+    InputIteratorT d_in,
+    OutputIteratorT d_out,
+    int num_segments,
+    BeginOffsetIteratorT d_begin_offsets,
+    EndOffsetIteratorT d_end_offsets,
+    ReductionOpT reduction_op,
+    T initial_value,
+    cudaStream_t stream = nullptr)
   {
     // Integer type for global offsets
     using OffsetT = detail::common_iterator_value_t<BeginOffsetIteratorT, EndOffsetIteratorT>;
+    using integral_offset_check = ::cuda::std::is_integral<OffsetT>;
 
-    return DispatchSegmentedReduce<InputIteratorT,
-                                   OutputIteratorT,
-                                   BeginOffsetIteratorT,
-                                   EndOffsetIteratorT,
-                                   OffsetT,
-                                   ReductionOpT>::Dispatch(d_temp_storage,
-                                                           temp_storage_bytes,
-                                                           d_in,
-                                                           d_out,
-                                                           num_segments,
-                                                           d_begin_offsets,
-                                                           d_end_offsets,
-                                                           reduction_op,
-                                                           initial_value,
-                                                           stream);
+    static_assert(integral_offset_check::value, "Offset iterator type should be integral.");
+
+    return custom_segmented_reduce<InputIteratorT, OutputIteratorT, BeginOffsetIteratorT, EndOffsetIteratorT, OffsetT>(
+      integral_offset_check{}, // feels like integral_offset_check::type{}
```

Comment:
I am confused by this. 

We have a static assert in front so there should never be a case where this is passed with `false_type`

Reply:
@miscco I'm not sure what you mean. Static asserts don't prohibit instantiations of subsequent templates: https://godbolt.org/z/Td1Y35f9r.  There's a case when false type is passed. It happens when user accidentally misuse API. The tag dispatch approach hides pages of underlying errors and instead leads to a single error coming from the front static assert.

Reply:
Did I totally miss the `false_type` declaration directly above.

I do not believe this is a question for this PR and I definitely wont block it, but I am not sure whether tag dispatch paired with an unimplemented method is a better fit than traditional SNIFAE / concepts.